url594c007908_hugobraga_College_tgz-pJ8-main_GraphJ8/src/algorithms/SIM.java

```diff
@@ -0,0 +1,355 @@
+			ArrayList<Group> groups = setMSCGroups(unmarked, uncoveredV2, false);
+			if (uncoveredV2.size() > 0) {
+				groups.addAll(setV1ToCoverNextTer(marked, informedSet, uncoveredV2));
+			}
+			
+			//22/03 - It seems like this function is only important to set chosenItem
+			procDistDom(groups, v2Set);
+			
+	    	//adding new edges to the arborescence
+	    	addingTreesForDomSet(uninformedTerminals);
+	    	
+	    	updateInformedSet(quasiArborescence, source);
+	    	
+	    	//updating marked and unmarked set
+	    	Iterator<PseudoNode> markedIt = realChosenPseudo.iterator();
+	    	while (markedIt.hasNext()) {
+	    		PseudoNode ps = markedIt.next();
+	    		int value = ps.getInformed().getId();
+	    		marked.add(value);
+	    	}
+	    	unmarked.addAll(informedSet);
+	    	unmarked.removeAll(marked);
+	    	
+	    	//updating sortedTerminals and uninformedAdjl lists by eliminating nodes
+	    	for (int j = 0; j < size; j++) {
+	    		if ((quasiArborescence.getList(j) != null) //if there is a list for j in quasiArborescence then this node is in the final arborescence and it is not an uninformed terminal anymore  
+	    				|| (uninformedTerminals.contains(new Integer(j)))) { //if j was one of the next terminals to be removed, so it is not an uninformed terminal anymore
+	    			uninformedAdjl.deleteNode(j);
+	    		}
+	    	}
+	    	//deleting terminals which were covered even they were not members of nextList
+	    	deleteCoveredTerminals();
+	    	
+        	uninformedAdjMatrix.refillMatrix();
+        	uninformedAdjMatrix.minCostPaths();	
+        	
+        	if (!flagViolatedTerminals) {
+        		flagViolatedTerminals = true;
+        		setViolatedTerminals();
+        	}
+		}
+		
+		return true;
+	}
+	
+	public void executeSPT() {
+		arborescenceSPT = buildSPT2Terminals(globalAdjMatrix, globalAdjl, terminals);
+	}
+	
+	public int getSPTMaxOutDegree() {
+		return getMaxOutDegree(arborescenceSPT);		
+	}
+	
+	public double[] getSPTCosts() {
+		return getCosts(globalAdjMatrix, terminals);
+	}	
+	
+	public int getArbMaxOutDegree() {
+		return getMaxOutDegree(quasiArborescence);
+	}
+	
+	
+	private void setFinalMatrix(AdjList arb) {
+		if (!finalMatrix) {
+			uninformedAdjMatrix.fillMatrixFromAdjList(arb);
+			uninformedAdjMatrix.minCostPaths();	
+			finalMatrix = true;
+		}
+	}	
+	
+	public double getTotalTerminalsCVR(double minCVR[], double maxCVR[]) {
+		setFinalMatrix(quasiArborescence);
+		
+		double terFinalCosts[] = getCosts(uninformedAdjMatrix, terminals);
+		double terSpannerCosts[] = getSpannerQuant();
+		//double min, max;
+		
+		minCVR[0] = 0;
+		maxCVR[0] = 0;
+		double ratios = 0;
+		for (int i = 0; i < terFinalCosts.length; i++) {
+			double val = terFinalCosts[i]/terSpannerCosts[i];
+			if (val > maxCVR[0])
+				maxCVR[0] = val;
+			if ((val < minCVR[0]) || (minCVR[0] == 0))
+				minCVR[0] = val;
+			ratios += terFinalCosts[i]/terSpannerCosts[i];
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 236:243, 261:268, 290:298.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.